### PR TITLE
Instance: Fix deadlock during failed snapshot creation

### LINKED ISSUE #13466: Creating a snapshot of a ZFS backed container with 0 bytes free results in hung lxd when doing snapshot operations
Ubuntu Noble
LXD 5.21.1 LTS

Creating an LXD container on a zfs backed filesystem, where you've set a quota (we set refquota flag on the pool as well, not sure if it matters) and then completely fill up the disk with dd - where profile disk is set to:

`size: 450GB`

And df shows
```
Filesystem              Size  Used Avail Use% Mounted on
lxd/containers/bigdata  420G  xxxG   xxxM xxx% /
```

If we then do

`dd if=/dev/zero of=temp.bin bs=1G count=420`

And make sure df shows 0 byes available, and on the host zfs list also shows 0 bytes available

And then do 

`lxc snapshot --reuse --no-expiry bigdata mysnapshot`

LXD will hang forever. You should see the command just sitting there if you do `ps aux`. What's worse, if you kill the snapshot operation, and other operations like snapshot delete will also hang. Only remedy was to do `snap restart lxd` and then lxd perked up immediately, we could free some space and redo the snapshot (which worked fine as soon as some space was free on disk).

Just snapshotting with zfs works instantly, so I suspect LXD is trying to write some data to the instance and this is what's hanging. How much space free on a zfs volume is required for a snapshot to work?

### PR DESCRIPTION
**Pull request by MggMuggins**
Fixes #13466

We need a non-locking delete on an `Instance`; otherwise a failure while mounting the instance/updating the backup file will deadlock.

I don't love the type assertion here. However, since `CreateInternal` returns `Instance`, we don't have a lot of options beyond enriching that interface or providing an additional one.

In order to avoid using type assertions we'd need to define an internal interface `instance` for `delete` and possibly a few other functions in the drivers/ package and implement it for `lxc`/`qemu`. Alternatively, moving the `Interface` interface into the same package as the drivers would allow us to define some private methods on `Interface` instead.

(Updating the backup file on the full volume was very likely the failure in #13466; this is after applying the fix):
```
$ lxc snapshot lucky-bulldog
Error: Failed to create file "/var/lib/lxd/containers/microcloud-test_lucky-bulldog/backup.yaml": open /var/lib/lxd/containers/microcloud-test_lucky-bulldog/backup.yaml: disk quota exceeded
```

I will throw a snapshot failure test on here this afternoon.

LXD-1117

### CODE REVIEW

File: lxd/instance/drivers/driver_common.go [hunk 1]
```diff
@@ -727,7 +727,16 @@ func (d *common) snapshotCommon(inst instance.Instance, name string, expiry time
...
 
-	revert.Add(func() { _ = snap.Delete(true) })
+	revert.Add(func() {
+		switch s := snap.(type) {
+		case *lxc:
+			_ = s.delete(true)
+		case *qemu:
+			_ = s.delete(true)
+		default:
+			panic("Delete snapshot of instance type other than lxc or qemu")
```
Comment (tomponline) [inline lxd/instance/drivers/driver_common.go:737]: we dont panic the daemon please

File: test/suites/snapshots.sh [hunk 2]
```diff
@@ -415,3 +415,29 @@ test_snap_volume_db_recovery() {
...
+
+test_snap_fail() {
+  local lxd_backend
+  lxd_backend=$(storage_backend "$LXD_DIR")
+
+  ensure_import_testimage
+
+  if [ "${lxd_backend}" = "zfs" ]; then
+    # Containers should fail to snapshot when root is full (can't write to backup.yaml)
+    lxc launch testimage c1 --device root,size=3GiB
```
Comment (MggMuggins) [inline test/suites/snapshots.sh:427]: `testimage` is very small; decrease the size here

File: test/suites/snapshots.sh [hunk 3]
```diff
@@ -415,3 +415,29 @@ test_snap_volume_db_recovery() {
...
+    lxc exec c1 -- dd if=/dev/urandom of=/root/big.bin count=100 bs=100M || true
+
+    exitCode=0
+    timeout 30s lxc snapshot c1 || exitCode=$?
```
Comment (MggMuggins) [inline test/suites/snapshots.sh:431]: `lxc` is already wrapped in `timeout 120` so skip this & the err code check

File: test/suites/snapshots.sh [hunk 4]
```diff
@@ -415,3 +415,29 @@ test_snap_volume_db_recovery() {
...
+
+    lxc ls
+
+    # If the timeout was reached, probably a deadlock
+    # https://github.com/canonical/lxd/issues/13466
+    [ "${exitCode}" -ne 124 ]
+
+    [ "$(lxc ls -c nS -f csv | awk -F ',' '/c1/{print $2}')" -eq 0 ]
+
+    lxc delete c1
```
Comment (MggMuggins) [inline test/suites/snapshots.sh:441]: `-f`

File: test/suites/snapshots.sh [hunk 5]
```diff
@@ -415,3 +415,29 @@ test_snap_volume_db_recovery() {
...
   lxc delete -f c1
 }
+
+test_snap_fail() {
+  local lxd_backend
+  lxd_backend=$(storage_backend "$LXD_DIR")
+
+  ensure_import_testimage
+
+  if [ "${lxd_backend}" = "zfs" ]; then
```
Comment (MggMuggins) [inline test/suites/snapshots.sh:425]: Try for other backends

Comment (MggMuggins) [reply test/suites/snapshots.sh:425]: See comment below

**Comment by MggMuggins** [comment]
`lxc stop` also fails on zfs when the disk is full:
`Error: Stopping the instance failed: Failed clearing ownership: chown /home/wesley/Workspace/lxd/test/tmp.KFM/iaA/containers/c1: disk quota exceeded`

The container does become `STOPPED`, but this prevents `lxc delete -f c1` from removing the container, hence the separate call to `lxc stop`. I'm happy to open another bug for this if it's undesired; it's inconvenient but easy to work around.

I tried `rm /root/big.bin` before deleting the instance and while I was able to remove the file, the dataset `USED`/`AVAIL` percentages didn't change. I doubt that there was a process still holding the file open (no `lsof` in testimage but only two processes, `init` and `ps` for `lxc exec c1 -- ps`) so I'm sure this is some zfs feature/behavior I'm not considering/don't know about.

Other backends:
| Backend | Result |
|---|---|
| `dir` | :x: :`lxc snapshot c1` succeeds |
| `btrfs` | :x: `lxc snapshot c1` succeeds; 3 MiB root disk required |
| `lvm` | :x: `lxc snapshot c1` succeeds; 9 MiB root disk required | 

Didn't try Ceph as I don't have a cluster handy

Let me know if you'd like other changes; thanks!

File: test/suites/snapshots.sh [hunk 6]
```diff
@@ -415,3 +415,25 @@ test_snap_volume_db_recovery() {
...
+    # Containers should fail to snapshot when root is full (can't write to backup.yaml)
+    lxc launch testimage c1 --device root,size=2MiB
+    lxc exec c1 -- dd if=/dev/urandom of=/root/big.bin count=100 bs=100K || true
+
+    ! lxc snapshot c1 || false
+
+    [ "$(lxc ls --columns nS --format csv | awk --field-separator , '/c1/{print $2}')" -eq 0 ]
+
+    # Stopping a container where / is a full dataset fails
+    lxc stop c1 || true
```
Comment (tomponline) [inline test/suites/snapshots.sh:435]: Please use `lxc stop -f c1` to do a forceful shutdown

File: test/suites/snapshots.sh [hunk 7]
```diff
@@ -415,3 +415,25 @@ test_snap_volume_db_recovery() {
...
+  ensure_import_testimage
+
+  if [ "${lxd_backend}" = "zfs" ]; then
+    # Containers should fail to snapshot when root is full (can't write to backup.yaml)
+    lxc launch testimage c1 --device root,size=2MiB
+    lxc exec c1 -- dd if=/dev/urandom of=/root/big.bin count=100 bs=100K || true
+
+    ! lxc snapshot c1 || false
+
+    [ "$(lxc ls --columns nS --format csv | awk --field-separator , '/c1/{print $2}')" -eq 0 ]
```
Comment (tomponline) [inline test/suites/snapshots.sh:432]: Please add a comment explaining what this does

**Comment by tomponline** [comment]
> `lxc stop` also fails on zfs when the disk is full:
> `Error: Stopping the instance failed: Failed clearing ownership: chown /home/wesley/Workspace/lxd/test/tmp.KFM/iaA/containers/c1: disk quota exceeded`

I think we should log a warning but continue the onstop hook function if the error contains "disk quota exceeded" (or if there is a better way to detect that error).

**Comment by MggMuggins** [comment]
Correct me if I'm wrong, but the failure of those chown/chmod would only leave the mount point owned by a uid that could potentially be reused for another container. All that does is allow the mount point to be [traversed](https://github.com/canonical/lxd/blob/main/lxd/instance/drivers/driver_lxc.go#L2263) by a process running under that uid _that has access to the parent dir_. Agreed, not a concern. Thanks for your patience.

**Comment by tomponline** [comment]
> Correct me if I'm wrong, but the failure of those chown/chmod would only leave the mount point owned by a uid that could potentially be reused for another container. All that does is allow the mount point to be [traversed](https://github.com/canonical/lxd/blob/main/lxd/instance/drivers/driver_lxc.go#L2263) by a process running under that uid _that has access to the parent dir_. Agreed, not a concern. Thanks for your patience.

Yes thats my understanding, and either way that doesn't succeeded whether we return an error or not, its just that by returning an error further bad things happen because the instance's devices are not cleaned up on the host side.

**Review by tomponline — APPROVED** [review]

**Pull request merged**